### PR TITLE
fix `partially_inline!` with `isdefined`

### PR DESCRIPTION
**Pull request by simeonschaub**
ref https://github.com/JuliaLang/julia/pull/40562#issuecomment-824912109

### CODE REVIEW

**Comment by vtjnash** [comment]
This is still not fully correct (since it first generates invalid IR, then tries to fix cases it can detect as such), but I think it may be valid for any cases that dynamically reach here?

**Comment by simeonschaub** [comment]
We could assert that the `isdefined` expression contains one of either a slot number, global ref, symbol or static parameter before we do the replacements, but that means the IR was already invalid before. If the original IR and replacements are valid, I don't see any way this would produce anything invalid.

**Comment by vtjnash** [comment]
Typically if there is a replacement made (unless it is a renumbering), that implies that the value was defined, which is what generally makes this valid dynamically. But it might therefore depend on the given semantics of the inlining.

**Comment by simeonschaub** [comment]
Ah, I see what you mean. I can't really think of any examples where this wouldn't be the case. The only packages using this function directly seem to be Cassette and IRTools (https://juliahub.com/ui/CodeSearch?q=partially_inline%21&u=all&t=all), so this is probably fine.

**Comment by JeffBezanson** [comment]
Just as a matter of style, I think it would be better to implement this by adding a case for `isdefined` expressions just like all other forms that need to be handled. That might seem like it repeats some logic for SlotNumber, but overall I think it's a better structure.

File: base/meta.jl [hunk 1]
```diff
@@ -403,6 +403,27 @@ function _partially_inline!(@nospecialize(x), slot_replacements::Vector{Any},
...
+                    replacement = slot_replacements[id]
+                    if isa(replacement, Union{Core.SlotNumber, GlobalRef, Symbol})
+                        return Expr(:isdefined, replacement)
+                    else
+                        @assert !isa(replacement, Expr)
+                        return true
+                    end
+                end
+                return Expr(:isdefined, Core.SlotNumber(id + slot_offset))
+            elseif isexpr(arg, :static_parameter)
```
[the remark this answers is not in the document]
Comment (simeonschaub) [reply base/meta.jl:421]: That's a good point, but for `partially_inline!` to work, you would still need to provide a value for the static parameter. Perhaps we should allow the sparam replacements to have undefined entries and only replace with true if the replacement is defined?

[the remark this answers is not in the document]
Comment (simeonschaub) [reply base/meta.jl:424]: Does this look good to you?

**Comment by simeonschaub** [comment]
Bump.

**Comment by simeonschaub** [comment]
Friendly bump :)

File: base/meta.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (JeffBezanson) [reply base/meta.jl:424]: We typically use TypeVars to indicate an unknown static parameter, but this function doesn't seem to support that (yet) so we can leave that for now, and this looks good.

**Review by JeffBezanson — APPROVED** [review]

**Comment by JeffBezanson** [comment]
Needs rebase

**Pull request merged**